File: src-tauri/src/commands/hub.rs
//! Pointing this desktop at a `fleet-hub`, and pointing it back.
//!
//! Three commands, all of them about **this process** rather than about a
//! fleet, which is why all three behave identically in both modes (their rows
//! in `backend::verdicts::VERDICTS` say so, each with its reason):
//! `hub_status` reports which fleet this window is onto, `hub_pair` redeems a
//! pairing code, and `hub_disconnect` forgets the pairing.
//!
//! # Disconnect does not revoke
//!
//! Clearing the token here removes it from **this machine**. The client row
//! stays on the hub and the token stays valid there until an operator revokes
//! it (`fleet-hub client revoke`), and a paired client is refused
//! `revoke_client` by design, so this app could not do it even if it wanted
//! to. Settings says so in those words; anything softer would leave someone
//! believing a stolen laptop had been locked out when it had not.
//!
//! # The mode is decided once, at startup
//!
//! [`crate::backend::Backend::resolve`] runs in `lib.rs`'s setup closure and
//! is deliberately never re-read — a mode flip mid-run would leave half the
//! app talking to a hub and half to the local store. So pairing and
//! disconnecting change what the *next* launch will do, and
//! [`HubStatus::restart_required`] is how Settings knows to say so rather
//! than leaving the user wondering why nothing happened.

use crate::backend::pairing::{PairTransport, TcpPairTransport};
use crate::backend::token_store::TokenStore;
use crate::backend::Backend;
use fleet_core::ipc_error::IpcError;
use fleet_core::store::Store;
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};
use tauri::State;

/// Which fleet this window is onto, as the frontend sees it.
///
/// Carries no token and must never carry one: this is a Tauri return value,
/// so it is serialised into the webview, where it would show up in a devtools
/// network panel and in any crash report the webview makes.
#[derive(Serialize, Debug, Clone, PartialEq, Eq, Default)]
pub struct HubStatus {
    /// What this process is doing **right now**.
    pub remote: bool,
    /// The hub it is a window onto; `None` when standalone.
    pub url: Option<String>,
    pub client_name: Option<String>,
    /// `full` or `readonly` as the hub recorded it at pairing time, when this
    /// process has just paired. `None` after a restart — the mode is the
    /// hub's to know, and nothing stores it.
    pub client_mode: Option<String>,
    /// `hub.remote_url` as **stored**, which differs from [`Self::url`]
    /// between a pairing and the restart that applies it.
    pub configured_url: Option<String>,
    pub configured_client_name: Option<String>,
    /// `hub.client_plaintext_token`
    /// ([`crate::backend::ALLOW_PLAINTEXT_KEY`]): this operator decided to
    /// send the client token over plain http to a routable host. The wire
    /// name stays `allow_plaintext` — it is a field between this app's
    /// halves, not the stored key, and not the daemon's setting of that name.
    pub allow_plaintext: bool,
    /// Why a configured hub is not in use, or why the one in use is risky —
    /// for the stored configuration, as the next launch would resolve it.
    /// The same string `Backend::resolve` logs — it belongs in front of a
    /// person, not only in a log file.
    pub warning: Option<String>,
    /// The stored configuration no longer matches the running mode.
    pub restart_required: bool,
    /// Set when a hub is configured but this launch could not use it, with
    /// the reason. This process then owns NOTHING — no reconcile tick, no
    /// usage poll, no control API — and refuses every fleet command.
    pub unavailable: Option<String>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct HubPairArgs {
    /// The hub's base URL as typed. Normalised here.
    pub url: String,
    /// The 8-character code from `fleet-hub pair --name <this machine>`.
    pub code: String,
    /// The user has seen [`fleet_core::ipc_error::codes::E_HUB_PLAINTEXT`]
    /// and chose to send the token in the clear anyway.
    #[serde(default)]
    pub allow_plaintext: bool,
}

#[tauri::command]
pub fn hub_status(
    backend: State<'_, Backend>,
    store: State<'_, Arc<Mutex<Store>>>,
    tokens: State<'_, Arc<dyn TokenStore>>,
) -> Result<HubStatus, IpcError> {
    logic::status(&backend, &store, tokens.inner().as_ref())
}

#[tauri::command]
pub async fn hub_pair(
    args: HubPairArgs,
    backend: State<'_, Backend>,
    store: State<'_, Arc<Mutex<Store>>>,
    tokens: State<'_, Arc<dyn TokenStore>>,
) -> Result<HubStatus, IpcError> {
    logic::pair(
        &TcpPairTransport,
        &backend,
        &store,
        tokens.inner().as_ref(),
        args,
    )
    .await
}

/// Whether this window's live link to the hub is up, for a window that
/// mounted after the last `hub:connection` event. See `backend::connection`.
#[tauri::command]
pub fn hub_connection(
    status: State<'_, Arc<crate::backend::connection::HubConnectionStatus>>,
) -> crate::backend::connection::HubConnection {
    status.current()
}

/// Whether a client token is sitting on this machine with **no hub
/// configured** — a credential nothing reads and, until this existed, nothing
/// offered to clear.
///
/// # The state this exists for
///
/// Pairing writes two stores that no transaction spans. Since `1e14414` the
/// token is written **last**, so a crash can no longer strand one; before it,
/// the token went first, and a crash between the two writes left a live
/// fleet-wide bearer token beside a blank `hub.remote_url`. An install that
/// took that crash carries the token across the upgrade.
///
/// [`Backend::resolve`] deliberately does not look for it: a blank URL is
/// standalone, and a keychain query on every launch is the call that prompts
/// or blocks on a locked macOS keychain (see
/// [`crate::backend::UnavailableHub`]). The question is asked here instead —
/// once, when a person opens Settings, which is both the moment a keychain
/// prompt is explicable and the moment the answer can be acted on: Settings
/// shows Disconnect, and [`hub_disconnect`] clears the token.
///
/// Answers `false` whenever a hub **is** configured. The token is not stranded
/// then — it belongs to that hub, and Settings already offers Disconnect for
/// the working and the [`crate::backend::Backend::Unavailable`] case alike.
///
/// The token never leaves the token store: this returns a bool.
#[tauri::command]
pub fn hub_stranded_token(
    store: State<'_, Arc<Mutex<Store>>>,
    tokens: State<'_, Arc<dyn TokenStore>>,
) -> Result<bool, IpcError> {
    logic::stranded_token(&store, tokens.inner().as_ref())
}

/// Forget the pairing. **Revokes nothing** — see the module docs.
#[tauri::command]
pub fn hub_disconnect(
    backend: State<'_, Backend>,
    store: State<'_, Arc<Mutex<Store>>>,
    tokens: State<'_, Arc<dyn TokenStore>>,
) -> Result<HubStatus, IpcError> {
    logic::disconnect(&backend, &store, tokens.inner().as_ref())
}

pub(crate) mod logic {
    use super::*;
    use crate::backend::{
        normalise_base_url, pairing, plaintext_risk, ALLOW_PLAINTEXT_KEY, CLIENT_NAME_KEY,
        REMOTE_URL_KEY,
    };
    use fleet_core::ipc_error::{codes, lock};

    /// Re-resolving here rather than reusing the startup decision is the
    /// point: `backend` says what this process is *doing*, and a fresh
    /// resolution says what the stored settings *ask for*. The difference
    /// between them is exactly [`HubStatus::restart_required`], and the
    /// resolution's warning is the reason a configured hub is not in use.
    pub fn status(
        backend: &Backend,
        store: &Mutex<Store>,
        tokens: &dyn TokenStore,
    ) -> Result<HubStatus, IpcError> {
        let stored = Backend::resolve_detail(store, tokens);
        // Read for display only. A failure here is not worth failing the
        // whole status over — the warning above already explains a store
        // that cannot be read.
        let (configured_url, configured_client_name, allow_plaintext) = match store.lock() {
            Ok(s) => (
                setting(&s, REMOTE_URL_KEY),
                setting(&s, CLIENT_NAME_KEY),
                matches!(
                    setting(&s, ALLOW_PLAINTEXT_KEY).as_deref(),
                    Some("true" | "1" | "yes")
                ),
            ),
            Err(_) => (None, None, false),
        };
        Ok(HubStatus {
            remote: backend.is_remote(),
            url: backend.remote().map(|c| c.base_url.clone()),
            client_name: backend.remote().map(|c| c.client_name.clone()),
            // Only a pairing in this session knows the mode; a restart
            // forgets it, and guessing `full` would be a claim about the
            // hub's records that this app cannot make.
            client_mode: None,
            configured_url,
            configured_client_name,
            allow_plaintext,
            warning: stored.warning,
            restart_required: &stored.backend != backend,
            // What THIS process is doing, like `remote`: a hub fixed since
            // launch is still unavailable here until the restart, which
            // `restart_required` says.
            unavailable: backend.unavailable().map(|hub| hub.reason.clone()),
        })
    }

    /// A stored setting, treating blank as absent — `hub_disconnect` writes
    /// `""` rather than deleting, because the settings registry has no
    /// delete.
    fn setting(s: &Store, key: &str) -> Option<String> {
        s.get_setting(key)
            .ok()
            .flatten()
            .map(|v| v.trim().to_string())
            .filter(|v| !v.is_empty())
    }

    pub async fn pair(
        transport: &dyn PairTransport,
        backend: &Backend,
        store: &Mutex<Store>,
        tokens: &dyn TokenStore,
        args: HubPairArgs,
    ) -> Result<HubStatus, IpcError> {
        // Everything that can be refused without spending the code is
        // refused first. A pairing code dies on first use and the hub allows
        // about ten attempts a minute from one address, so making someone
        // burn one to be told their URL has a typo is a real cost.
        let base_url = normalise_base_url(args.url.trim()).map_err(|why| {
            IpcError::new(
                codes::E_INVALID,
                format!(
                    "{:?} is not a usable hub address ({why}) — it needs a scheme \
                     and a host, e.g. https://fleet.example.com",
                    args.url.trim()
                ),
            )
        })?;
        let code = args.code.trim().to_string();
        if code.is_empty() {
            return Err(IpcError::new(
                codes::E_INVALID,
                "enter the pairing code from `fleet-hub pair --name <this machine>`",
            ));
        }
        // The client half of the hub's own `--allow-plaintext`, asked here
        // because this is the first moment a URL exists and the LAST moment
        // before a fleet-wide credential is minted for it.
        let risk = plaintext_risk(&base_url);
        if let Some(risk) = &risk {
            if !args.allow_plaintext {
                return Err(IpcError::new(
                    codes::E_HUB_PLAINTEXT,
                    format!(
                        "{risk}. Use https://, or pair again with \"send it in the clear \
                         anyway\" if this hop really is already private (a tunnel, a VPN, \
                         a container network)."
                    ),
                ));
            }
        }

        let client = pairing::redeem(transport, &base_url, &code).await?;

        // From here the code is spent: it cannot be presented twice.
        //
        // A pairing writes two stores that no transaction spans — the token
        // in the keychain or the 0600 file, the settings in `state.db` — so a
        // crash can land between any two writes, and every state in between
        // is one a launch may start from. The order makes each of them safe:
        //
        // 1. Forget any previous hub's token. From here until step 3 there is
        //    no token at all, and a configured URL with no token resolves
        //    `Unavailable`: it owns nothing, says why, and offers Disconnect.
        // 2. Write the settings, the URL last.
        // 3. Store the new token — beside the URL of the hub that issued it.
        //
        // The old order stored the token first. A crash then left it beside
        // the previous hub's URL, and the next launch presented hub B's
        // bearer token to hub A; or beside no URL at all, where no launch
        // reads it and no screen offers to clear it.
        let previous = read_settings(store);
        tokens.clear().map_err(|e| {
            IpcError::new(
                codes::E_IO,
                format!(
                    "paired with {base_url}, but this machine's secure storage would not \
                     forget the previous token ({e}), so nothing was changed here — the \
                     code is spent, so mint a fresh one on the hub and try again"
                ),
            )
        })?;
        if let Err(e) = write_settings(store, &base_url, &client.name, args.allow_plaintext) {
            // Best effort: the store that just failed may fail again, and
            // with no token stored every state it leaves is safe anyway.
            let _ = restore_settings(store, &previous);
            return Err(IpcError::new(
                codes::E_IO,
                format!(
                    "paired with {base_url}, but the settings could not be saved ({}) — \
                     no token was kept. If this app was paired with a hub before, it now \
                     has no token for that one either, and the next launch will say so. \
                     The code is spent, so mint a fresh one on the hub and try again.",
                    e.message
                ),
            ));
        }
        if let Err(e) = tokens.set(&client.token) {
            // Put the settings back: a first pairing leaves no trace, and a
            // re-pairing leaves the previous hub configured with no token,
            // which the next launch reports rather than acts on.
            let _ = restore_settings(store, &previous);
            return Err(IpcError::new(
                codes::E_IO,
                format!(
                    "paired with {base_url}, but this machine's secure storage refused \
                     the token ({e}), so the settings were put back and nothing was kept \
                     — the code is spent, so mint a fresh one on the hub and try again"
                ),
            ));
        }

        let mut out = status(backend, store, tokens)?;
        // The one moment this app ever knows its access level: the hub said
        // so in the pairing response, and nothing stores it.
        out.client_mode = Some(client.mode);
        Ok(out)
    }

    /// The URL is written LAST: until it is, the next launch still sees
    /// the previous URL (or none), and with no token stored — `pair` has
    /// cleared it — either is safe.
    fn write_settings(
        store: &Mutex<Store>,
        base_url: &str,
        client_name: &str,
        allow_plaintext: bool,
    ) -> Result<(), IpcError> {
        let s = lock(store)?;
        s.set_setting(CLIENT_NAME_KEY, client_name)?;
        s.set_setting(
            ALLOW_PLAINTEXT_KEY,
            if allow_plaintext { "true" } else { "false" },
        )?;
        s.set_setting(REMOTE_URL_KEY, base_url)?;
        Ok(())
    }

    /// The three hub settings as stored before a pairing, so a failed one
    /// can put them back. Absent reads as blank, which is what `resolve`
    /// already treats as "no hub".
    fn read_settings(store: &Mutex<Store>) -> [(&'static str, String); 3] {
        let read = |key| {
            store
                .lock()
                .ok()
                .and_then(|s| s.get_setting(key).ok().flatten())
                .unwrap_or_default()
        };
        [
            (CLIENT_NAME_KEY, read(CLIENT_NAME_KEY)),
            (ALLOW_PLAINTEXT_KEY, read(ALLOW_PLAINTEXT_KEY)),
            (REMOTE_URL_KEY, read(REMOTE_URL_KEY)),
        ]
    }

    fn restore_settings(
        store: &Mutex<Store>,
        previous: &[(&'static str, String); 3],
    ) -> Result<(), IpcError> {
        let s = lock(store)?;
        for (key, value) in previous {
            s.set_setting(key, value)?;
        }
        Ok(())
    }

    /// See [`super::hub_stranded_token`]. The settings are consulted FIRST so
    /// that a configured hub — the case Settings already handles — never
    /// reaches the token store at all; the keychain is asked only in the state
    /// this is about.
    pub fn stranded_token(store: &Mutex<Store>, tokens: &dyn TokenStore) -> Result<bool, IpcError> {
        // A settings store that cannot be read is not evidence of a stranded
        // token: `Backend::resolve` already answers that case loudly
        // (`unreadable_settings`), and answering `true` here would put a
        // "leftover credential" warning in front of someone whose actual
        // problem is a database it also names. Not stranded, as far as this
        // can tell.
        let configured = lock(store)?
            .get_setting(REMOTE_URL_KEY)?
            .is_some_and(|v| !v.trim().is_empty());
        if configured {
            return Ok(false);
        }
        // The message is the token store's own diagnostic text, which never
        // contains the secret (`TokenStore`'s contract). A failure is
        // reported rather than swallowed, but the caller is free to stay
        // quiet: a keychain that will not open is not evidence of a leftover.
        tokens.get().map(|t| t.is_some()).map_err(|e| {
            IpcError::new(
                codes::E_IO,
                format!("could not check this machine's secure storage for a leftover client token: {e}"),
            )
        })
    }

    /// Forget the pairing on **this machine**. Revokes nothing on the hub —
    /// see the module docs, and the wording Settings uses.
    ///
    /// The token goes first: if the settings write then fails, what is left
    /// is a URL with no token, which [`Backend::resolve`] already handles
    /// loudly ("configured but no client token is stored"). The other order
    /// would leave a live token against a hub the app no longer names, which
    /// nothing reports.
    pub fn disconnect(
        backend: &Backend,
        store: &Mutex<Store>,
        tokens: &dyn TokenStore,
    ) -> Result<HubStatus, IpcError> {
        tokens.clear().map_err(|e| {
            IpcError::new(
                codes::E_IO,
                format!("this machine's secure storage refused to forget the token: {e}"),
            )
        })?;
        {
            let s = lock(store)?;
            // The settings registry stores strings and has no delete; blank
            // is what `resolve` already treats as "no hub".
            s.set_setting(REMOTE_URL_KEY, "")?;
            s.set_setting(CLIENT_NAME_KEY, "")?;
            s.set_setting(ALLOW_PLAINTEXT_KEY, "")?;
        }
        status(backend, store, tokens)
    }
}

#[cfg(test)]
#[path = "tests_hub.rs"]
mod tests;

File: src/lib/hub.ts
// Which fleet this window is onto, and what that costs.
//
// Standalone the desktop *is* the fleet: its own database, its own SSH
// connections, its own reconcile tick, its own control API. Paired with a
// `fleet-hub` it is a client of someone else's fleet, and three families of
// thing stop working from here:
//
//   1. **Fleet administration.** The desktop pairs as an ordinary client, and
//      a client is never the fleet's administrator: adding, removing and
//      hiding hosts, provisioning, asset sync and secrets are the hub's.
//   2. **Things about THIS machine.** The terminal's PTY, the local
//      prerequisites check, SSH tunnels, the asset-catalog git checkout.
//   3. **PARITY OR REFUSAL.** A mutation routes to the hub only where the
//      desktop's arguments map one-to-one onto the tool's parameters. Where
//      they do not — `repair_session` with `explicit: false`, the automatic
//      pre-attach check — the command refuses rather than routing, because
//      routing would have quietly turned it into the tool's own (always
//      explicit) repair. A refusal is visible; a dropped field is not.
//      Nobody should later "fix" a refusal like this by wiring a lossy
//      mapping.
//
// The backend returns `E_LOCAL_ONLY` for these, with a message naming where
// the operation does work. This module is the *other* half: the reason in
// front of the control, before the click, so that a paired desktop looks like
// a paired desktop rather than like a broken one.
import { get, writable } from 'svelte/store';
import { invokeCmd, type IpcError, type Result } from './result';
import { hubConnection, type HubConnection } from './hub_connection';

/** Mirrors the backend `HubStatus` (`src-tauri/src/commands/hub.rs`). */
export interface HubStatus {
  /** What this process is doing right now. */
  remote: boolean;
  /** The hub it is a window onto; null when standalone. */
  url: string | null;
  client_name: string | null;
  /** `full` | `readonly`, known only in the session that paired. */
  client_mode: string | null;
  /** `hub.remote_url` as stored — differs from `url` until a restart. */
  configured_url: string | null;
  configured_client_name: string | null;
  allow_plaintext: boolean;
  /** Why a configured hub is not in use, or why the one in use is risky. */
  warning: string | null;
  /** The stored configuration no longer matches the running mode. */
  restart_required: boolean;
  /**
   * Set when a hub is configured but THIS launch could not use it (no stored
   * token, a keychain that would not open, plain http without the opt-in, a
   * URL that does not parse), with the reason. The backend then owns nothing
   * — no reconcile tick, no usage poll, no control API — and refuses every
   * fleet command, rather than quietly becoming a second brain for the hub's
   * fleet. `remote` is false in this state: nothing is talking to a hub.
   */
  unavailable: string | null;
}

export const STANDALONE: HubStatus = {
  remote: false,
  url: null,
  client_name: null,
  client_mode: null,
  configured_url: null,
  configured_client_name: null,
  allow_plaintext: false,
  warning: null,
  restart_required: false,
  unavailable: null,
};

/**
 * The last status the backend gave us. Standalone until told otherwise, which
 * is the safe default in one direction only — it enables things — so
 * `loadHubStatus` runs before the first render that can act on it, and a
 * FAILED reload deliberately keeps the previous value rather than falling
 * back to standalone.
 */
export const hubStatus = writable<HubStatus>({ ...STANDALONE });

export async function loadHubStatus(): Promise<Result<HubStatus>> {
  const r = await invokeCmd<HubStatus>('hub_status');
  if (r.ok && r.value) hubStatus.set(r.value);
  return r;
}

/** Redeem a pairing code. Takes effect at the next launch — the backend mode
 *  is resolved once, at startup, on purpose. */
export async function hubPair(
  url: string,
  code: string,
  allowPlaintext: boolean,
): Promise<Result<HubStatus>> {
  const r = await invokeCmd<HubStatus>('hub_pair', {
    args: { url: url.trim(), code: code.trim(), allow_plaintext: allowPlaintext },
  });
  if (r.ok && r.value) hubStatus.set(r.value);
  return r;
}

/**
 * Forget the pairing on this machine.
 *
 * This **revokes nothing**. The client row stays on the hub and the token
 * stays valid there until an operator revokes it (`fleet-hub client revoke`),
 * and a paired client is refused `revoke_client` by design — so the app could
 * not revoke even if it wanted to. Every string this module and the Settings
 * dialog show says so; anything softer would leave someone believing a stolen
 * laptop had been locked out when it had not.
 */
export async function hubDisconnect(): Promise<Result<HubStatus>> {
  const r = await invokeCmd<HubStatus>('hub_disconnect');
  if (r.ok && r.value) hubStatus.set(r.value);
  return r;
}

/**
 * Whether a client token is sitting on this machine with **no hub
 * configured** — a fleet-wide credential no launch reads.
 *
 * Pairing used to store the token before the URL, so a crash between the two
 * writes stranded one. That order is fixed (the token goes last now), but an
 * install that took the crash carries the token across the upgrade, and
 * `hub_status` never finds it: a blank URL is standalone, and the backend
 * deliberately does not query the OS keychain on every launch just in case.
 *
 * So Settings asks once, on open — the one moment a keychain prompt is
 * explicable — and offers {@link hubDisconnect}, which clears it. Deliberately
 * **not** a field of {@link HubStatus}: it is a one-off question about a legacy
 * state, not part of which fleet this window is onto, and nothing branches on
 * it but the Hub section.
 */
export async function hubStrandedToken(): Promise<Result<boolean>> {
  return invokeCmd<boolean>('hub_stranded_token');
}

// ---- what a hub client cannot do from here ---------------------------------

/**
 * The half of each reason that is specific to the action. `hubBlock` adds the
 * hub's name, so every sentence ends up naming both what and where.
 *
 * The keys are the command names wherever there is one, so a reader can find
 * the matching `backend.local_only(...)` in `src-tauri/src/commands/`.
 */
const REASONS = {
  // --- fleet administration: refused to a client by design -----------------
  add_host:
    'this desktop is a paired client, and a client is never the fleet’s administrator',
  remove_host:
    'this desktop is a paired client, and a client is never the fleet’s administrator',
  hide_host:
    'this desktop is a paired client, and a client is never the fleet’s administrator',
  discover_hosts:
    'host discovery reads this machine’s SSH config, and a paired client does not administer the fleet anyway',
  provision_hosts:
    'provisioning rewrites every host’s hook block to report to whichever app ran it, and this desktop is a paired client',
  apply_sync:
    'asset sync writes to the hosts over SSH, which a paired client does not do',
  set_secret:
    'sync secrets belong to the machine that runs the sync, and this desktop is a paired client',
  install_fleet_hook:
    'the hook points at a control API this app is not running, because a paired client runs none',
  host_tokens:
    'these are a fleet owner’s per-host tokens; this desktop is a paired client and has none',
  mcp_configure:
    'starting a second control API against a fleet that already has one is the failure remote mode exists to prevent',
  // Trackers (work graph M3): the hub's work_admin is master-only.
  add_tracker:
    'trackers and their credentials are fleet administration, and a client is never the fleet’s administrator — use `fleet-hub tracker add`',
  update_tracker:
    'trackers and their credentials are fleet administration, and a client is never the fleet’s administrator — use `fleet-hub tracker`',
  set_tracker_credential:
    'tracker credentials live on the hub, and a client is never the fleet’s administrator — use `fleet-hub tracker set-credential`',
  test_tracker:
    'the hub tests its trackers with its own credentials, and a client is never the fleet’s administrator — use `fleet-hub tracker test`',
  remove_tracker:
    'trackers and their credentials are fleet administration, and a client is never the fleet’s administrator — use `fleet-hub tracker remove`',
  tracker_sync_metrics:
    'sync metrics live in the hub’s memory, and a client is never the fleet’s administrator — use `fleet-hub tracker status`',
  // Organisations (work graph M5): the per-host tokens' boundary is set on
  // the hub only; `work_admin` is master-only.
  add_org:
    'organisations are the hosts’ security boundary, and a client is never the fleet’s administrator — use `fleet-hub org add`',
  update_org:
    'organisations are the hosts’ security boundary, and a client is never the fleet’s administrator — use `fleet-hub org set`',
  remove_org:
    'organisations are the hosts’ security boundary, and a client is never the fleet’s administrator — use `fleet-hub org rm`',
  add_org_rule:
    'which org a session belongs to is fleet administration, and a client is never the fleet’s administrator — use `fleet-hub org rule add`',
  remove_org_rule:
    'which org a session belongs to is fleet administration, and a client is never the fleet’s administrator — use `fleet-hub org rule rm`',
  assign_host_org:
    'a host’s org is its token’s boundary, set only by the fleet’s administrator — use `fleet-hub org assign-host`',
  assign_tracker_org:
    'which org a tracker belongs to is fleet administration, and a client is never the fleet’s administrator — use `fleet-hub org assign-tracker`',

  // --- things about THIS machine ------------------------------------------
  // (No `terminal` key: the terminal is not blocked by being a hub client.
  // `pty_open` spawns this machine's own `ssh`/`tmux`, so a paired desktop
  // attaches exactly as a standalone one does. The one session it cannot
  // attach is one on an agent host, and that is not the hub's doing — nothing
  // anywhere has an SSH route to it — so TerminalView says that itself rather
  // than through a reason ending "Do it on the hub".)
  check_local_prereqs:
    'the setup checklist is about running a fleet from this machine, which the hub is doing instead',
  tunnel_status: 'the tunnels belong to whichever process owns the fleet',
  mcp_status: 'this app runs no embedded control API while a hub owns the fleet',
  catalog_config:
    'the asset catalog is a git checkout on the machine that owns the fleet; the hub serves its asset list to any paired client (list_assets), but not the configuration and checkout this panel is built on',
  get_fleet_settings:
    'these settings drive the reconcile tick, the GC sweeper and the playbooks, which the hub runs and this app does not',
  list_account_usage:
    'this app does not poll account usage while a hub owns the fleet, so its cache stays empty',
  refresh_account_usage:
    'it reads the account’s usage over this machine’s SSH connection to the host',
  set_account_nickname:
    'the nickname lives in the hub’s database and there is no tool to set it',

  // --- acted out over this machine's SSH, and the hub has no tool for it ---
  repo_write:
    'the hub exposes no git-write tool — a remote client must not stage or commit under a running agent; do it in the session',
  add_project:
    'it clones or adopts a checkout using this machine’s SSH and GitHub credentials',
  purge_project:
    'it deletes Claude Code state on every host over this machine’s SSH connections, and the hub exposes no tool for it',
  inspect_safe_kill:
    'it inspects the worktree over this machine’s SSH connection, and the hub exposes no tool for it',
  discard_kill_session:
    'the hub exposes no tool that discards a worktree and kills in one step — use Safe remove’s "Ask Claude" path, or do it from the hub',
  dismiss_agent_session:
    'use Kill instead: the hub’s kill_session removes an inactive agent from the list exactly as this would',

} as const;

export type HubAction = keyof typeof REASONS;

/** Every action this module has a reason for — the list the tests sweep. */
export const HUB_ACTIONS = Object.keys(REASONS) as HubAction[];

/**
 * Whether this process owns its fleet — standalone, and not pointed at a hub,
 * working or not. Everything that fetches or shows a fleet-owner-only panel
 * asks this rather than `!status.remote`, because a configured hub this launch
 * cannot use is not a hub client and is not standalone either: it owns
 * nothing.
 */
export function ownsTheFleet(status: HubStatus = get(hubStatus)): boolean {
  return !status.remote && !status.unavailable;
}

/**
 * F1's sentence, shared by `hubBlock` and `hubActionBlocked`: a hub is
 * configured but THIS launch could not use it, so the backend owns no fleet
 * at all and refuses every command — routed or refused — the same way.
 */
function unavailableReason(status: HubStatus): string | null {
  if (!status.unavailable) return null;
  // Not "do it on the hub": the hub is the problem, and every action is
  // refused until it is fixed.
  return `Not available: ${status.unavailable}. This app is set to use that hub, so it manages no fleet of its own until that is fixed — pair again, or Disconnect, in Settings → Hub.`;
}

/**
 * Why `action` is unavailable from this window, or `null` when it is
 * available. `null` in standalone mode, always: nothing here may change what
 * a standalone app does.
 */
export function hubBlock(action: HubAction, status: HubStatus = get(hubStatus)): string | null {
  const unavailable = unavailableReason(status);
  if (unavailable) return unavailable;
  if (!status.remote) return null;
  const where = status.url ?? 'the hub';
  return `${REASONS[action]}. Do it on the hub (${where}).`;
}

// ---- offline gating: routed mutations while the live link is down ---------

/**
 * Action keys for the mutations that ROUTE to the hub — the ones a paired
 * client can still send, as long as the live connection to it is up. This is
 * a SUBSET of the backend's routed commands (`src-tauri/src/backend/
 * verdicts.rs`): the mutations the UI actually gates on the connection
 * state, not every routed command — a routed read like `list_sessions` has
 * no control to disable and so is not here.
 *
 * `hub_verdicts.test.ts` holds this list (and `REASONS` below) accountable
 * to `hub_verdicts.generated.json`, the JSON generated straight from
 * `VERDICTS`, rather than this file importing that JSON and deriving from
 * it: keeping the literal here lets `RoutedAction` stay exactly the union
 * type it is today. The names usually match the `#[tauri::command]` fn name;
 * the one place they don't is `set_friendly_name`, whose command is
 * `set_session_friendly_name` (the tool it routes to is `set_friendly_name`,
 * which is what `tests_routing.rs` names the case after, and what the
 * cross-check test maps through too).
 */
export const ROUTED_ACTIONS = [
  'send_prompt',
  'kill_session',
  'safe_kill_session',
  'rename_session',
  'set_friendly_name',
  'restart_session',
  'spawn_review',
  'recreate_session',
  'dismiss_ghost_session',
  'new_bg_session',
  'delete_worktree',
  'cancel_task',
  'probe_host',
  'move_session',
  'new_session',
  'repair_session',
  'link_session_work',
  'reject_session_work',
  'unlink_session_work',
  'confirm_session_work',
  'set_work_project_trust',
  'start_work',
  'start_work_multi',
  'request_work_handover',
  'tidy_apply',
  'archive_session_work',
  'unarchive_session_work',
  'snooze_tidy',
  'never_tidy',
  'dismiss_reopened',
  'name_session_work',
  'rename_work_item',
] as const;

export type RoutedAction = (typeof ROUTED_ACTIONS)[number];

const ROUTED_ACTION_SET: ReadonlySet<string> = new Set(ROUTED_ACTIONS);

/** The offline banner's sentence, reworded as a per-control reason: what a
 *  click would need instead of what the whole window is missing. `null` for
 *  `standalone` / `connected`, which never block a routed action. */
function offlineReason(conn: HubConnection, url: string | null): string | null {
  const hub = url ?? 'the hub';
  switch (conn.state) {
    case 'connecting':
      return `Still connecting to ${hub} — try again once it's connected.`;
    case 'reconnecting':
      return `${hub} is unreachable right now (reconnecting, attempt ${conn.attempt}) — try again once it's back.`;
    case 'offline':
      return `${hub} is unreachable right now (retrying, attempt ${conn.attempt}) — try again once it's back.`;
    case 'hub_too_old':
      return `${hub}'s version is incompatible with this app — its wire contract (revision ${conn.hub_contract}) is older than the ${conn.min_contract} this app requires. Update the hub.`;
    case 'hub_too_new':
      return `${hub}'s version is incompatible with this app — its wire contract (revision ${conn.hub_contract}) is newer than the ${conn.max_contract} this app understands. Update this app.`;
    default:
      return null;
  }
}

/**
 * The one place a control checks BOTH halves of "can I send this": refusal
 * (`hubBlock`, `REASONS`) and the live connection (`$hubConnection`), so a
 * call site is a one-liner and the two checks cannot drift apart.
 *
 * Refusal wins over offline: an action the hub never accepts from a client
 * says so, not "try again once connected" (which it never will be, and would
 * send someone chasing a connection that was never the problem). `standalone`
 * — no hub configured — never blocks anything, in either half; a key that is
 * neither refused nor routed (reads, navigation, and the handful of commands
 * that run the same in both modes) is never blocked here either.
 */
export function hubActionBlocked(
  action: HubAction | RoutedAction,
  status: HubStatus = get(hubStatus),
  conn: HubConnection = get(hubConnection),
): string | null {
  if (Object.hasOwn(REASONS, action)) return hubBlock(action as HubAction, status);
  if (!ROUTED_ACTION_SET.has(action)) return null;
  const unavailable = unavailableReason(status);
  if (unavailable) return unavailable;
  if (!status.remote) return null;
  if (conn.state === 'connected') return null;
  return offlineReason(conn, status.url);
}

/**
 * The next step a hub client needs for an error that does not carry one.
 *
 * `E_CONFIRM_REQUIRED` is the trap this exists for. With
 * `mcp.confirm_destructive` on, the hub refuses `kill_session`,
 * `delete_worktree`, `move_session`, `cancel_task` and `repair_session` until
 * someone approves them — and the desktop's own confirmation dialog answers
 * **this process's** queue, which in remote mode is always empty. So the
 * click is refused, the dialog never appears, and without this sentence
 * there is nothing anywhere saying that the approval has to happen on the
 * hub.
 *
 * "This window will follow" is a consequence of the event bridge, not a hope:
 * once the operator approves, the hub emits `session:killed` /
 * `worktree:removed` / `task:updated` like any other change and the bridge
 * re-emits it here. So the sentence must not say "then refresh".
 */
export function hubNextStep(
  error: IpcError,
  status: HubStatus = get(hubStatus),
): string | null {
  if (!status.remote) return null;
  const where = status.url ?? 'the hub';
  switch (error.code) {
    case 'E_CONFIRM_REQUIRED':
      return `${where} is holding this until someone confirms it there, and this desktop's confirmation dialog answers only its own queue, which is empty. Approve it on the hub — this window will follow.`;
    case 'E_UNAUTHORIZED':
      return `${where} no longer accepts this desktop's client token. Pair again in Settings → Hub.`;
    default:
      return null;
  }
}
